File: rogue-server/data/player_data.py
import config

from collections import deque

from check.auto_clicker_check import AutoClickerCheck
from check.packet_check import PacketCheck

from tracker.trackers.action_tracker import ActionTracker
from tracker.trackers.collision_tracker import CollisionTracker
from tracker.trackers.movement_tracker import MovementTracker
from tracker.trackers.ping_trackcer import PingTracker
from tracker.trackers.potion_tracker import PotionTracker


class PlayerData:

    def __init__(self, id):
        self.checks = {}
        self.id = id

        self.violations = deque()
        self.ticks_existed = 0
        self.entity_id = 0

        self.action_tracker = ActionTracker(self)
        self.ping_tracker = PingTracker(self)
        self.potion_tracker = PotionTracker(self)
        self.movement_tracker = MovementTracker(self)
        self.collision_tracker = CollisionTracker(self)

        self.trackers = [
            self.action_tracker,
            self.ping_tracker,
            self.potion_tracker,
            self.movement_tracker,
            self.collision_tracker
        ]

    def has_violations(self):
        return len(self.violations) > 0

    def get_and_pop_violations(self):
        violations = [len(self.violations)]
        for i in range(len(self.violations)):
            violations.append(self.violations[i])
        self.violations.clear()

        return violations

    def add_violation(self, check_name, check_type, violations):
        data = config.get_check_data(check_name, check_type)
        self.violations.append({
            'action': 'ban' if violations > data['max_violations'] else data['action'],
            'checkName': check_name,
            'checkType': check_type,
            'violations': violations,
            'maxViolations': data['max_violations']
        })

    def handle_packet(self, event):
        packet = event['packet']

        if event['type'] == 'in_flying':
            self.ticks_existed += 1
            self.entity_id = event['entityId']

        for tracker in self.trackers:
            tracker.handle(event)

        for check in self.checks:
            if isinstance(check, PacketCheck) or isinstance(check, AutoClickerCheck):
                check.handle(event)

File: rogue-server/src/main/java/org/hostile/anticheat/data/PlayerData.java
package org.hostile.anticheat.data;

import com.google.gson.JsonArray;
import com.google.gson.JsonObject;
import lombok.Getter;
import org.hostile.anticheat.AntiCheatServer;
import org.hostile.anticheat.check.type.Check;
import org.hostile.anticheat.check.type.impl.packet.PacketCheck;
import org.hostile.anticheat.event.PacketEvent;
import org.hostile.anticheat.packet.inbound.WrappedPacketPlayInFlying;
import org.hostile.anticheat.tracker.impl.entity.EntityTracker;
import org.hostile.anticheat.tracker.impl.movement.CollisionTracker;
import org.hostile.anticheat.tracker.impl.movement.MovementTracker;
import org.hostile.anticheat.tracker.impl.player.ActionTracker;
import org.hostile.anticheat.tracker.impl.player.PingTracker;
import org.hostile.anticheat.tracker.impl.player.PotionTracker;

import java.util.ArrayList;
import java.util.List;
import java.util.Queue;
import java.util.UUID;
import java.util.concurrent.ConcurrentLinkedQueue;
import java.util.function.Predicate;
import java.util.stream.Collectors;

@Getter
public class PlayerData {

    private final List<Check<?>> checks = new ArrayList<>();
    private final Queue<JsonObject> violations = new ConcurrentLinkedQueue<>();

    private final UUID uuid;

    private final ActionTracker actionTracker;
    private final CollisionTracker collisionTracker;
    private final EntityTracker entityTracker;
    private final MovementTracker movementTracker;
    private final PingTracker pingTracker;
    private final PotionTracker potionTracker;

    private int entityId;
    private int ticksExisted;

    private String gameMode;
    private String worldName;

    public PlayerData(UUID uuid) {
        this.uuid = uuid;

        this.actionTracker = new ActionTracker(this);
        this.collisionTracker = new CollisionTracker(this);
        this.entityTracker = new EntityTracker(this);
        this.movementTracker = new MovementTracker(this);
        this.pingTracker = new PingTracker(this);
        this.potionTracker = new PotionTracker(this);

        this.checks.addAll(AntiCheatServer.getInstance().getCheckManager().getChecks(this));
    }

    public void handle(PacketEvent event) {
        if (event.getPacket() instanceof WrappedPacketPlayInFlying) {
            ++ticksExisted;

            JsonObject object = event.getJsonObject();

            entityId = object.get("entityId").getAsInt();
            gameMode = object.get("gamemode").getAsString();
            worldName = object.get("world").getAsString();
        }

        actionTracker.handle(event);
        collisionTracker.handle(event);
        entityTracker.handle(event);
        movementTracker.handle(event);
        pingTracker.handle(event);
        potionTracker.handle(event);

        checks.stream()
                .filter(check -> check instanceof PacketCheck)
                .forEach(check -> ((PacketCheck) check).handle(event));
    }

    public JsonArray getViolations() {
        JsonArray array = new JsonArray();

        violations.forEach(array::add);
        violations.clear();

        return array;
    }

    public void addViolation(JsonObject jsonObject) {
        this.violations.add(jsonObject);
    }

    @SuppressWarnings("unchecked")
    public <T extends Check<?>> List<T> getChecks(Predicate<Check<?>> predicate) {
        return checks.stream()
                .filter(predicate)
                .map(check -> (T) check)
                .collect(Collectors.toList());
    }
}